hail/python/hail/table.py

```diff
@@ -2544,8 +2544,43 @@ def to_matrix_table(self, row_key, col_key, row_fields=[], col_fields=[], n_part
         if len(col_key) == 0:
             raise ValueError(f"'to_matrix_table': require at least one col key field")
 
-        return hl.MatrixTable(TableToMatrixTable(
-            self._tir, row_key, col_key, row_fields, col_fields, n_partitions))
+        ht = self.key_by()
+
+        non_entry_fields = set(itertools.chain(row_key, col_key, row_fields, col_fields))
+        entry_fields = [x for x in ht.row if x not in non_entry_fields]
+
+        if not entry_fields:
+            raise ValueError(...)
+
+        col_data = hl.rbind(
+            hl.array(
+                hl.dict(
+                    ht.aggregate(
+                        hl.agg.collect_as_set(
+                            (ht.row.select(*col_key),
+                             ht.row.select(*col_fields)))))),
```

Comment:
In the (hopefully rare) case that some col keys are repeated with different col fields, this will collect all the copies. Would a `group_by` col key with a `take(1)` of col fields perform better?

Ah, I see that's what you do for rows. I assume there's something about the `group_by` aggregator that makes it worse than collecting all different fields values for each key?

Reply:
yeah, good idea.